### DISCUSSION TITLE
Best way to operate on an arbitrary part of an array?

### DISCUSSION THREAD

**Discussion by Scratch-net** [Consensys/gnark#1301] [Q&A]
Hey there!

So I've got a circuit which does block cipher decryption and the result is a fixed length array of decrypted data (bytes, bits, does not matter)
Next I need to process a part of that data (say, hash it) to prove that "a hash of some data indeed exists in this encrypted data from indexes A to B"

I'm ok with the "data" to be exactly 31 bytes every time to fit into BN254 element, I can use bitmasks for excluding redundant part. But can't get my head around the way to "take" these bytes from an arbitrary part of a block, say "starting from position 10".
Is there any way around it?

**Comment by Scratch-net** [comment]
Basically I need .FromBinary() where digits can be specified as frontend.Variable slice indices

**Reply by ivokub** [reply]
There is no exactly a direct way - you would have to decompose the full BN254 element into bits (using [`std/math/bits`](https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/math/bits#ToBinary)) and then construct bytes as needed. If you're not interested in the lower bits for example then you could use the [`std/math/bitslice`](https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/math/bitslice#Partition) package to split the BN254 into two parts and then do the corresponding bitwise decomposition on the upper part.

**Reply by Scratch-net** [reply]
Unfortunately neither my input (512 or 1024 bits) nor its subset would fit into a single element. It might work with bitslice though if I figure out how to reassemble everything in a constant number of steps. Thanks for the advice

**Reply by ivokub** [reply]
Usually in that case when I need to do SHA2 etc. I use `std/math/uints` package which has wrappers for U32 and U64. And then I provide the input as `[]uints.U32` or `[]uints.U8`. See for example https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/hash#BinaryHasher

**Reply by Scratch-net** [reply]
As we cannot use conditional loops in gnark we need to code everything in constant number of steps. So if one user needs to hash "10 bytes from pos 50" and the other "hash 60 bytes from pos 2" it becomes problematic.
I've done it through a hint for now but I'd rather not
Also, "split" parameter in `Partition()` is `uint`, not `frontend.Variable`

**Reply by ivokub** [reply]
I see. Yes, in this case `std/math/bitslice` doesn't work. And I don't think we have a nice built-in tool for it neither.

How I would do it is to have a mask which is set to 1 for bytes I want to include in the hash and the assert that mask is set correctly (first location of mask is the given initial pos; sum of masks is total number of bytes you want to hash; every mask bit is 0-1). Have also look at the https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/hash#BinaryFixedLengthHasher interface which allows to hash variable-length inputs. Our `sha2` hasher [implements it](https://pkg.go.dev/github.com/consensys/gnark@v0.11.0/std/hash/sha2#New), but would be possible to extend to sha3 also.

**Reply by Scratch-net** [reply]
Thanks, will definitely take a look!
